Add a general generator for icon package maintainers

mallardduck opened the pull request:
This PR adds the ability for `blade-icons` consumer libraries to easily create a generation script. The `GenerateCommandBuilder` accomplishes this by providing a fluent API to build a `SingleCommandApplication` that functions in 3 phases:

1. Generate temp files for normalization from source files,
2. Normalize the SVGs in the temp directory, (optional, only used when closure provided)
3. Relocate the normalized SVGs to their destination path.

Using a builder style pattern seems to be a better fit for both end-users and providing a good general purpose solution. At the current state I think it should be much more functional for most use cases compared to the first pass attempt.

However, this is still a rather basic script and it could have some short comings for some Icon packs. Namely if the icon pack stores the "default" (or regular) set in the root directory and has variant sets in sub directories.

For those cases it may make more sense to introduce a `IconSetParameters` DTO of sorts that would encapsulate the icon sets state and location. More input from other authors, or direct investigation into other icon packs may be needed.

---

Here's what my boxicons package would use to make the generator work:

```php
GenerateCommandBuilder::create('blade-boxicons')
    ->fromNpmPackage('boxicons') // Optionally specify an `npm` package to load from
    ->fromSourceSvgDirectory('/svg') // Specify a directory, if an npm package isn't set, this can be anywhere
    ->withIconSets([
        IconSetConfig::create('regular')
            ->setInputFilePrefix('bx-'),
        IconSetConfig::create('logos')
            ->setInputFilePrefix('bxl-'),
        IconSetConfig::create('solid')
            ->setInputFilePrefix('bxs-'),
    ])
    ->withSvgNormalisation(function (string $tempFilepath, string $iconSet) {
        $doc = new DOMDocument();
        $doc->load($tempFilepath);
        /**
         * @var DOMElement $svgElement
         */
        $svgElement = $doc->getElementsByTagName('svg')[0];
        $svgElement->removeAttribute('width');
        $svgElement->removeAttribute('height');
        $svgElement->removeAttribute('viewBox');
        $svgElement->setAttribute('fill', 'currentColor');
        $svgElement->setAttribute('stroke', 'none');
        $svgElement->setAttribute('viewBox', '0 0 24 24');
        $doc->save($tempFilepath);

        $svgLine = trim(file($tempFilepath)[1]);
        file_put_contents($tempFilepath, $svgLine);
    })
    ->run();
```

---
# Example Implementations:
My Boxicons Package: https://github.com/mallardduck/blade-boxicons/pull/1/files
@owenvoke 's FontAwesome package: https://github.com/owenvoke/blade-fontawesome/pull/25
@driesvints Heroicons package: https://github.com/blade-ui-kit/blade-heroicons/pull/19

mallardduck committed:
Add abstract command to provide common foundation for a generator command

mallardduck committed:
update docs

mallardduck committed:
Add generator command examples

owenvoke commented:
I feel like this would be kind of cool to have as a [single-command application](https://symfony.com/doc/current/components/console/single_command_tool.html) in the packages, so you'd basically just have an executable `generate` file which contains:

```php
#!/usr/bin/env php
<?php

require __DIR__.'/vendor/autoload.php';

use BladeUI\Icons\Console\GenerateCommand;

GenerateCommand::create()
    ->fromNpmPackage('@fortawesome/fontawesome-free') // Optionally specify an `npm` package to load from
    ->fromSourceSvgDirectory('/svgs') // Specify a directory, if an npm package isn't set, this can be anywhere
    ->withIconSets([
        'regular' => 'far-',
        'brands' => 'fab-',
        'solid' => 'fas-',
    ])
    ->withSvgNormalisation(function (string $tempFilepath, string $iconSet) {
        // Any required SVG manipulation, called after copying the files (equivalent to normalizeSvgFile)
    })
    ->run();
```

This makes the process even cleaner, and doesn't require extending the class unless the package requires something additional. It also means there would only be a single file (the `generate` binary).

mallardduck commented:
@owenvoke - I'm pretty novice in the symfony space still -since I usually stick with laravel- so I'm not sure if what you're showing is possible. But from my limited playing around with it I know it's not impossible. And TBH that seems like a pretty decent end user API - one that I could even potentially see supporting both the "manually downloaded icon source" and the "fetches this from npm icon source" methods.

Seems like ultimately the main "issue" i'm thinking of with your example is just really semantics of naming - since I think it needs to be a `*CommandBuilder` class of sorts - potentially. Really appreciate that feedback on this PR, I'll start working towards something similar to the example you provided.

---

The one area I'm wondering about expanding that neither of our implementations cover yet is non-uniform directory structures. As it seems some icon packs have the "default" set at `svg/*.svg` and then other sets at `svg/{SET}/*.svg`. So that makes me consider promoting the icon set from a key/value pair to a config class. 🤔

mallardduck committed:
Refactor to a CommandBuilder partern

mallardduck committed:
Adjust constructor and command output

owenvoke commented:
Wow! I love the changes, looks great. 👏🏻

mallardduck committed:
Refactor to add IconSetConfig DTO and allow single folder output

mallardduck committed:
remove unused property

mallardduck committed:
ensure cross OS compatibility for all paths.

mallardduck commented:
Alright - so the current state of things seems to be pretty optimal for the icon packages I sent PRs to. I'm sure @driesvints may have some feedback directly, but also curious if we should seek feedback from other authors?

---

One thing I think we could improve potentially is around the variation found in different Icon packages source projects. I've done my best to encapsulate this logic in the `IconSetConfig`.

So in the case of Heroicons, multiple source folders/sets can render down to a single IconSet prefixes (and output folder) where each SVG file also has a filename prefix. And in my case with boxicons I can render them out to multiple sets with unique IconSet prefixes.

In theory this same functionality would be where someone could configure things to look for the "main" set in a "root" folder. And then secondary sets in sub folders.

mallardduck commented on src/Console/IconSetConfig.php:
```diff
@@ -0,0 +1,80 @@
+
+    public function getSourceFilePath(string $iconFileName): string
+    {
+        return $this->svgSourcePath . DIRECTORY_SEPARATOR . $this->name . DIRECTORY_SEPARATOR . $iconFileName;
```
This is where we'd need to make adjustments for Icon packs where the "main" set are in a root folder.

mallardduck replied:
This is also resolved by using laravel's `Filesystem::files()` method to get a list of files. Given that this method returns a list of `\Symfony\Component\Finder\SplFileInfo` I removed this method from `IconSetConfig` all together.

mallardduck commented on src/Console/GenerateCommandBuilder.php:
```diff
@@ -0,0 +1,180 @@
+     */
+    private function getDirectoryFileList(string $dirPath, IconSetConfig $iconSetConfig): array
+    {
+        return array_diff(scandir($dirPath), ['..', '.']);
```
For Icon packs with the "main" set in a root dir, the alternative sets would be in directories.
So to support those, we will need to expand this to remove all directories along with `.` and `..`.

mallardduck replied:
I think by using Laravel's `Filesystem::files()` we get the benefit of not worrying about this potential issue. So at the very least those icon sets won't be broken now.

owenvoke commented on src/Console/GenerateCommandBuilder.php:
```diff
@@ -0,0 +1,180 @@
+            }
+
+            // Copy to final destination...
+            copy($tempFile, $finalFile);
```
This appears to fail if the destination directory doesn't exist. 👍🏻 Probably worth having something like:
```suggestion
            mkdir(dirname($finalFile), 0777, true);
            copy($tempFile, $finalFile);
```

mallardduck replied:
Should be good now - used the `ensureDirExists` helper method.

mallardduck committed:
Ensure final destination directories exist too.

mallardduck committed:
update code styles

mallardduck committed:
more style fixes

mallardduck committed:
whitespace fixes

owenvoke commented:
Random request, could we add a `clearDestinationDirectory()` method, for my Font Awesome package they seem to remove icons due to copyright sometimes and it might be useful. Although, to be honest, I'm not keen on removing icons as it might be a breaking change to someone's site. 🤷🏻

mallardduck committed:
Add ability for GenerateCommandBuilder to clear output directory

mallardduck commented:
@owenvoke - yeah, no problem adding that at all. As you guessed, your fontawesome package lost the Adobe Icon when I ran it with that change.

mallardduck committed:
Refactor to reuse laravel Filesystem class and remove code

mallardduck committed:
tweak docblocks

mallardduck committed:
Ensure that file input/output prefix are both respected

owenvoke commented:
Any changes or help needed on this? 👍🏻 🙂

driesvints commented:
Sorry, haven't found time to look into this yet. I'll try to do this soon but no promises.

driesvints committed:
cleanup

driesvints commented:
Thanks for your patience. I finally found time to look into all this. This is a fantastic idea, great job on the PR so far!

I made some changes but cannot push to the PR. @mallardduck can you allow me to push to this pr as well as the one on the heroicons repo? Then you'll see the changes I already made. 

I haven't found time yet to look into everything like the IconSetConfig but I first wanted to pitch my own ideas. I don't like it that much that every package will need its own command. I believe we can do better here. What I want to propose is that each icon package that'll make use of icon generation will have a `config/generation.php` file with all the settings needed to do the icon generation. And that Blade Icons ships with the binary instead. So you can use it like this:

```zsh
./vendor/bin/generate-icons heroicons
```

Or maybe even without providing a name? We'd just need to make the icon generator read out the config file and do all the generation stuff. That way packages will only need to provide a config file. 

What do you think @owenvoke @mallardduck? Can we make that happen?

owenvoke commented:
I think that would definitely be possible. 👍🏻 And probably a quite nice way of doing it. I feel like the generation without the name would probably be best by default? So by default, it would publish all icon sets in the config, otherwise it would just publish the one named.

Also, not sure if we would want a less generic name for the generator script?

One thought, does it matter that the `config/generation.php` would be included by default in the released package? Or should we always add this to the `.gitattributes` as a file to ignore on export?

driesvints commented:
Btw @mallardduck please don't make changes to this PR and the heroicons one until I had a chance to push my own changes.

> I feel like the generation without the name would probably be best by default? So by default, it would publish all icon sets in the config, otherwise it would just publish the one named.

Sounds good

> Also, not sure if we would want a less generic name for the generator script?

Yeah definitely. Just couldn't figure out one.

> One thought, does it matter that the config/generation.php would be included by default in the released package? Or should we always add this to the .gitattributes as a file to ignore on export?

Don't feel that matters much. Packages can choose to add it to the .gitattributes. I'm planning on creating a package skeleton for new icon packages after this PR is merged. I'll add it to the gitattributes.

mallardduck commented:
@driesvints 🤔  On both of these PRs the "Allow edits by Maintainers" option is already checked. So you should already have access to push to those branches that the PRs source. Just for fun I unchecked and rechecked them - hopefully that gives you proper write access to push changes on these PRs.

mallardduck commented:
As far as the alternative idea so that blade-icons will ship a binary and icon-packs just provide a config that seems reasonable. Excited to see what your changes look like!

driesvints commented:
@mallardduck thanks. I managed to push to both forks this time.

mallardduck commented:
@driesvints - awesome. just to confirm, I can start working on some of those adjustments discussed right?

driesvints commented:
@mallardduck yes! Really appreciate that 🙂

mallardduck committed:
add generate bin script...
this allows package maintainers to simply make the config file that builds the command.

mallardduck committed:
fix styles

mallardduck commented:
@driesvints the bin script changes were surprisingly simple to get working. At least with the implementation that doesn't take any user input. The current bin script is only meant to be run from the packages root - not a laravel app consuming the icon pack. It does seem to work really well though.

Should be ready for final review!

driesvints commented:
@mallardduck awesome. I'm a bit out of time this week. Not sure when exactly I'll get back to this but hope so soon. Thanks for your work!

owenvoke commented on bin/blade-icons-generate:
```diff
@@ -0,0 +1,14 @@
+
+require $currentDir.'/vendor/autoload.php';
+
+$packageGenerateCommand = require $currentDir.'/config/generation.php';
```
I feel like it would probably be good to have a nicer error message than the default failure when a file in a `require` call isn't found. Maybe something like:

```suggestion
$configurationPath = "{$currentDir}/config/generation.php";

if (! file_exists($configurationPath)) {
    echo 'The Blade Icons generator configuration file could not be found.';
    echo "Please create one at: {$configurationPath}";
    exit(1);
}

$packageGenerateCommand = require $configurationPath;
```

---

I guess all this sort of the stuff (clean output/messages) can be tidied up before release though.

mallardduck replied:
That's a great idea @owenvoke - updated now.

mallardduck committed:
Validate that the config file exists and can be found

mallardduck committed:
fix typo and improve output printing

mallardduck commented on src/Generation/IconGenerator.php:
```diff
@@ -0,0 +1,216 @@
+
+    private bool $useSingleIconSet = false;
+
+    private bool $safe = false;
```
@owenvoke - I know this feature was a concern you brought up initially.
So i'm curious what your thoughts are on the default for this feature.

Currently without explicitly using `safe` the destination folder will be nuked before processing.
As you pointed out it'd be good to keep up with upstream removals, however they can be breaking changes.

driesvints replied:
I'll document this clearly to anyone using it. Personally I'd want to keep on pair with icon sets as removing of icons usually involve copyrights like @owenvoke indicates. You don't want those lingering in your set. Sure, could be a breaking change but an icon missing is less severe than your app not working anymore. Plus Blade Icons is very explicit about missing icons by throwing exceptions.

mallardduck replied:
@driesvints - sounds good to me!

I was wondering if I should edit the readme to document some of this stuff. I held off since there's quite a lot to cover, so was wondering if it'd make sense on the blade-ui-kit docs.

driesvints replied:
@mallardduck I'll document everything once this is merged. I could end up refactoring some more so don't want you to do needless work. Thanks

driesvints committed:
Simplify code

driesvints committed:
Remove temp stuff

driesvints committed:
Remove icon config class

driesvints committed:
More cleanup

driesvints committed:
last cleanup

driesvints committed:
update composer.json

driesvints committed:
wip

driesvints commented:
Okay, I sort of worked the entire day on this PR and had way too much fun refactoring it to what it is now 😅 

Basically I stripped it down a lot. All functionality still remains but is much simpler now. Took down the Pr from 330 lines to 99 lines of code. Take a look also at the other commit I made for Blade Heroicons: https://github.com/blade-ui-kit/blade-heroicons/pull/19

@owenvoke @mallardduck Bit too tired now to type a lengthy explanation. Just look over the PR and let me know if that all looks good or not 😅

mallardduck commented on bin/blade-icons-generate:
```diff
@@ -0,0 +1,21 @@
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\SingleCommandApplication as Command;
+
+require __DIR__.'/../vendor/autoload.php';
```
This is less accurate at finding the autoloader than my original method was. I'm now getting errors because of this change, see:

```bash
± |experimental-generator U:1 ✗| → ./vendor/bin/blade-icons-generate

Warning: require(/Users/danpock/GitProjects/blade-boxicons/vendor/blade-ui-kit/blade-icons/bin/../vendor/autoload.php): Failed to open stream: No such file or directory in /Users/danpock/GitProjects/blade-boxicons/vendor/blade-ui-kit/blade-icons/bin/blade-icons-generate on line 9
```

driesvints replied:
Did you run `composer install`?

mallardduck replied:
Yes. I did, but you maybe overlooking that this "bin" script shouldn't use the `vendor` folder in `blade-icons`. But should instead use the `vendor` folder from the project it's being required by. 

`__DIR__` is the directory of the file itself - so `blade-icons/bin` meaning the `/../` only ever gets us to `blade-icons/` and then to `vendor/`. In total, this transforms `blade-icons/bin` to `blade-icons/vendor/autoload.php`.

However when an end-user installs using composer composer this folder won't exist at all. So it should use that icon packs vendor folder since `blade-icons` will live there already. See here:
```
~/GitProjects/blade-boxicons
± |experimental-generator ✓| → ll vendor/blade-ui-kit/blade-icons/vendor
ls: vendor/blade-ui-kit/blade-icons/vendor: No such file or directory
```

Also - once the package is installed in a consumer icon pack the path will be different. So using `__DIR__` and multiple `/../` would require numbers traversing to parent diers. Since for instance in my `blade-boxicons` (`~/GitProjects/blade-boxicons`) package this code will exist at:
* `~/GitProjects/blade-boxicons/vendor/bin/blade-icons-generate` (symlink)
* `~/GitProjects/blade-boxicons/vendor/blade-ui-kit/blade-icons/bin/blade-icons-generate` (real)

But `__DIR__` always resolves to the true location; so in either case, the actual path would need to be `__DIR__ . '/../../../autoload.php`.

mallardduck replied:
TL;DR: The only reason I think this worked for you is because you did do `composer install` inside of your `blade-icons` folder. And you're using `path` instead of `vcs`, but once you switch to `vcs` (which is basically what composer does by default) this will break.

```
        {
            "type": "path",
            "url": "../blade-icons"
        }
```

So if you deleted the vendor folder from your `blade-icons` folder while you develop on `blade-heroicons` then it's more accurate to how composer will install things for real. Since normally `blade-heroicons/vendor/blade-ui-ki/blade-icons/vendor/` should (and will) never exist for the end users of this package.

mallardduck committed:
Adjust autoloader to work more consistently

mallardduck committed:
allow tracking bin folder

mallardduck commented on .gitattributes:
```diff
@@ -1,7 +1,6 @@
 * text=auto
 
 /.github export-ignore
-/bin export-ignore
```
Before this change I was getting:
```shell
    Skipped installation of bin bin/blade-icons-generate for package blade-ui-kit/blade-icons: file not found in package
```

driesvints replied:
Ah, that snuck in actually. Thanks for letting me know.

mallardduck commented on bin/blade-icons-generate:
```diff
@@ -6,7 +6,12 @@ use Symfony\Component\Console\Input\InputInterface;
 use Symfony\Component\Console\SingleCommandApplication as Command;
 
-require __DIR__.'/../vendor/autoload.php';
+$autoloadPath = getcwd() . "/vendor/autoload.php";
```
For me doing this version of the code it works fine with either VCS (realistic to composer installs) and path based repos. So I think this would make more sense in the end?

driesvints replied:
I don't understand why this is needed. Can you explain a bit more extensively?

mallardduck replied:
See the replies here:
https://github.com/blade-ui-kit/blade-icons/pull/144#discussion_r657953501
https://github.com/blade-ui-kit/blade-icons/pull/144#discussion_r657958061

The way you're locally developing things is skewing what files should and shouldn't exist for end users. If you were to use `vcs` instead of `path` repos in composer:

```
        {
            "type": "vcs",
            "url": "https://github.com/mallardduck/blade-icons"
        }
```

Then you can test only the changes you made to refactor with:
```
        "blade-ui-kit/blade-icons": "dev-generator-dries",
```

And compare to:
```
        "blade-ui-kit/blade-icons": "dev-abstract-generator-command",
```

mallardduck reviewed (commented):
Other than the few comments and changes I've made things seem to work as advertised.
Once I updated my config to the more simple array based config sets I get the same output.

See here for my packages implementation: https://github.com/mallardduck/blade-boxicons/pull/1/files

Overall looks good if these changes I made work for you guys.

mallardduck commented:
@owenvoke - I've updated the PR on your repo as well and tested it. All looks good.
The litmus test for success here (IMHO) was:

1.  deleting all of `resources/svg/`,
2. checking that change in,
3. running the builder, and
4. checking those changes-in to ensure no files were actually "changed" in the end.

So basically the `safe` flag but a little more manual and direct to be more through.
Doing that on both of our repos after updating the config show'd zero changes.

owenvoke commented on .gitattributes:
```diff
@@ -1,6 +1,7 @@
 * text=auto
 
 /.github export-ignore
+/bin export-ignore
```
Isn't it necessary to have this in the export, otherwise when installing it as a dependency it'd be impossible to run the `blade-icons-generate` binary? 🤔

owenvoke replied:
Oh, this change doesn't seem to exist. Commits seem to just be in a weird order. 😕

- cafad03743f337ba5f59621ae7c7ba23434832d8 - 12 hours ago
- 71039c58ed03f56e9cdbc7ec841dfc45b9b4f7ec - 1 hour ago

But the first seems to override this. Oh well, seems to all be fine in the diff.

driesvints replied:
huh, you're right. What's going on here 🤨

driesvints replied:
Nuked that commit

driesvints committed:
update autoload

driesvints commented:
@mallardduck thanks for the explanation. I copied the same use as the envoy binary now and it works as expected.

driesvints commented:
Thanks for this @mallardduck & @owenvoke! I'll write docs and release in a few.

driesvints commented:
https://github.com/blade-ui-kit/blade-icons#generating-icons
